### PR TITLE
Fixes for Inertial Guidance and Kappa Guidance

### PR DESCRIPTION
**Pull request by BillNyeTheIE**
- Fixed issues with manual and guard mode targeting using inertial targeting
- Improved performance of Kappa guidance against lower altitude targets
Note: probably a lot of issues that still require fixing

### CODE REVIEW

File: BDArmory/Radar/VesselRadarData.cs [hunk 1]
```diff
@@ -2099,7 +2099,7 @@ private void DrawDisplayedContacts()
                             if (weaponManager.selectedWeapon.GetWeaponClass() == WeaponClasses.Missile || weaponManager.selectedWeapon.GetWeaponClass() == WeaponClasses.SLW)
                             {
                                 MissileBase currMissile = weaponManager.CurrentMissile;
-                                if (currMissile && (currMissile.TargetingMode == MissileBase.TargetingModes.Radar || currMissile.TargetingMode == MissileBase.TargetingModes.Heat))
+                                if (currMissile && (currMissile.TargetingMode == MissileBase.TargetingModes.Radar || currMissile.TargetingMode == MissileBase.TargetingModes.Heat || currMissile.TargetingMode == MissileBase.TargetingModes.Inertial))
                                 {
```
Comment (SuicidalInsanity) [inline BDArmory/Radar/VesselRadarData.cs:2103]: Could add GPS missiles to this too, since those can use a radar lock in lieu of a targetpod for ranging.

File: BDArmory/Control/MissileFire.cs [hunk 2]
```diff
@@ -773,11 +777,21 @@ public void ToggleGuardMode()
...
+                                    if (GpsUpdateMax > 0 && scanSpeed < GpsUpdateMax) GpsUpdateMax = scanSpeed;
+                                    _radarsEnabled = true;
+                                }
+                                else
+                                {
+                                    rd.Current.EnableRadar();
+                                    float scanSpeed = rd.Current.directionalFieldOfView / rd.Current.scanRotationSpeed * 2;
+                                    if (GpsUpdateMax > 0 && scanSpeed < GpsUpdateMax) GpsUpdateMax = scanSpeed;
+                                    _sonarsEnabled = true;
+                                }
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:793]: Do we want both Active and Passive sonars activated at guard, or just Passive, with Active toggled when needed for targeting?

File: BDArmory/Control/MissileFire.cs [hunk 3]
```diff
@@ -5954,7 +6024,7 @@ bool SmartPickWeapon_EngagementEnvelope(TargetInfo target)
                             case (WeaponClasses.SLW):
                                 {
                                     MissileLauncher SLW = item.Current as MissileLauncher;
-                                    if (SLW.TargetingMode == MissileBase.TargetingModes.Radar && (!_radarsEnabled && !SLW.radarLOAL)) continue; //dont select RH missiles when no radar aboard
+                                    if (SLW.TargetingMode == MissileBase.TargetingModes.Radar && (!(_sonarsEnabled || _radarsEnabled) && !SLW.radarLOAL)) continue; //dont select RH missiles when no radar aboard
                                     if (SLW.TargetingMode == MissileBase.TargetingModes.Laser && targetingPods.Count <= 0) continue; //don't select LH missiles when no FLIR aboard
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:6028]: should be just _sonarsEnabled, so a surface radar doesn't permit torp launches; radar doesn't work underwater.

File: BDArmory/Control/MissileFire.cs [hunk 4]
```diff
@@ -6360,10 +6441,12 @@ bool CheckEngagementEnvelope(IBDWeapon weaponCandidate, float distanceToTarget,
                             fireFOV = mlauncher.missileTurret ? mlauncher.missileTurret.turret.yawRange : mlauncher.multiLauncher && mlauncher.multiLauncher.turret ? mlauncher.multiLauncher.turret.turret.yawRange : -1;
 
                             MissileLaunchParams dlz = MissileLaunchParams.GetDynamicLaunchParams(ml, targetVessel.Velocity(), targetVessel.transform.position, fireFOV,
-                                (ml.TargetingMode == MissileBase.TargetingModes.Laser && BDATargetManager.ActiveLasers.Count <= 0 || ml.TargetingMode == MissileBase.TargetingModes.Radar && !_radarsEnabled && !ml.radarLOAL));
+                                (ml.TargetingMode == MissileBase.TargetingModes.Laser && BDATargetManager.ActiveLasers.Count <= 0 ||
+                                ml.TargetingMode == MissileBase.TargetingModes.Radar && !(_radarsEnabled || _sonarsEnabled) && !ml.radarLOAL ||
+                                ml.TargetingMode == MissileBase.TargetingModes.Inertial && !(_radarsEnabled || _sonarsEnabled || _irstsEnabled)));
                             if (vessel.srfSpeed > ml.minLaunchSpeed && distanceToTarget < dlz.maxLaunchRange && distanceToTarget > dlz.minLaunchRange)
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:6447]: Why are we querying sonar status for above-water missiles?

File: BDArmory/Control/MissileFire.cs [hunk 5]
```diff
@@ -2447,52 +2462,97 @@ IEnumerator GuardMissileRoutine(Vessel targetVessel, MissileBase ml)
...
-                                float angle = 999;
-                                float turretStartTime = attemptStartTime;
-                                if (mlauncher.missileTurret)
+                                float attemptStartTime = Time.time;
+
+                                float attemptLockTime = Time.time;
+                                while (ml && vesselRadarData && (!vesselRadarData.locked || (vesselRadarData.lockedTargetData.vessel != targetVessel)) && Time.time - attemptLockTime < 2f)
                                 {
-                                    while (Time.time - turretStartTime < Mathf.Max(targetScanInterval / 2f, 2) && mlauncher && targetVessel && angle > mlauncher.missileTurret.fireFOV)
+                                    if (vesselRadarData.locked)
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:2480]: Part of the point/implementation of INS was that is just needed TWS to launch against a target; this should instead be something like 
```
if (vesselRadarData)
{
    INSTarget = vesselRadarData.detectedRadarTarget(targetVessel, this);
    if (!INSTarget.exists && irsts.Count > 0 && _irstsEnabled)
    {
       INSTarget = vesselRadarData.activeIRTarget(null, this);
    }
    if (vessel && INSTarget.exists && targetVessel)
    {
        Vector3 TargetLead = MissileGuidance.GetAirToAirFireSolution(ml, INSTarget.vessel.CoM,  INSTarget.vessel.Velocity());
        designatedINSCoords = VectorUtils.WorldPositionToGeoCoords(TargetLead, INSTarget.vessel.mainBody);
     }...
```

File: BDArmory/Control/MissileFire.cs [hunk 6]
```diff
@@ -6921,14 +7015,16 @@ public void SendTargetDataToMissile(MissileBase ml, bool clearHeat = true)
...
                         {
-                            ml.targetGPSCoords = designatedGPSCoords;
+                            ml.targetGPSCoords = designatedINSCoords;
                             ml.TargetAcquired = true;
                             validTarget = true;
+                            designatedINSCoords = Vector3d.zero;
                         }
                         else
                         {
+                            Vector3d targetCoords = Vector3.zero;
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:7027]: designatedINSCoords is already a vector3.zero here, why the new variable?

File: BDArmory/Control/MissileFire.cs [hunk 7]
```diff
@@ -7102,8 +7204,10 @@ void GuardMode()
...
                                     }
                                 }
-                                MissileLaunchParams dlz = MissileLaunchParams.GetDynamicLaunchParams(CurrentMissile, guardTarget.Velocity(), guardTarget.CoM, -1, (CurrentMissile.TargetingMode == MissileBase.TargetingModes.Laser
-                                    && BDATargetManager.ActiveLasers.Count <= 0 || CurrentMissile.TargetingMode == MissileBase.TargetingModes.Radar && !_radarsEnabled && !CurrentMissile.radarLOAL));
+                                MissileLaunchParams dlz = MissileLaunchParams.GetDynamicLaunchParams(CurrentMissile, guardTarget.Velocity(), guardTarget.CoM, -1,
+                                    (CurrentMissile.TargetingMode == MissileBase.TargetingModes.Laser && BDATargetManager.ActiveLasers.Count <= 0 ||
+                                    CurrentMissile.TargetingMode == MissileBase.TargetingModes.Radar && !(_radarsEnabled || _sonarsEnabled) && !CurrentMissile.radarLOAL ||
+                                    CurrentMissile.TargetingMode == MissileBase.TargetingModes.Inertial && !(_radarsEnabled || _sonarsEnabled || _irstsEnabled)));
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:7211]: might make sense to make a new bool and a switch(targetingMode){} segment above this that toggles the bool as needed to keep this manageable

File: BDArmory/Control/MissileFire.cs [hunk 8]
```diff
@@ -7861,7 +7965,7 @@ public void PointDefenseTurretFiring()
                             if (PDMslTgts[MissileID].Vessel.Splashed && !missile.torpedo) viableTarget = false;
                             //need to see if missile is turreted (and is a unique turret we haven't seen yet); if so, check if target is within traverse, else see if target is within boresight
                             bool turreted = false;
-                            if (missile.TargetingMode == MissileBase.TargetingModes.Radar && _radarsEnabled && !missile.radarLOAL && MaxradarLocks < vesselRadarData.GetLockedTargets().Count) continue; //don't have available radar lock, move to next missile                            
+                            if (missile.TargetingMode == MissileBase.TargetingModes.Radar && (_radarsEnabled || _sonarsEnabled) && !missile.radarLOAL && MaxradarLocks < vesselRadarData.GetLockedTargets().Count) continue; //don't have available radar lock, move to next missile                            
                             MissileTurret mT = null;
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:7969]: this is going to cause problems on anything that has both radar AAMs and sonar torp interceptors, would need to be
``if (missile.TargetingMode == MissileBase.TargetingModes.Radar && (missile.torpedo ? _sonarsEnabled : _radarsEnabled) && !missile.radarLOAL && MaxradarLocks < vesselRadarData.GetLockedTargets().Count) continue; //don't have available radar lock, move to next missile``

File: BDArmory/Control/MissileFire.cs [hunk 9]
```diff
@@ -6960,20 +7092,21 @@ public void SendTargetDataToMissile(MissileBase ml, bool clearHeat = true)
...
                                     targetVessel = vesselRadarData.lockedTargetData.targetData.vessel;
+                                    validTarget = true;
                                 }
                                 else if (irsts.Count > 0) //or brightest ping on IRST
                                 {
-                                    validTarget = true;
                                     targetVessel = vesselRadarData.activeIRTarget(null, this).vessel;
+                                    validTarget = targetVessel;
                                 }
+                                vesselRadarData.LastMissile = ml;
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:7103]: Shouldn't this be inside 7093's if block? No need for this to be set if no radar lock.

File: BDArmory/Control/MissileFire.cs [hunk 10]
```diff
@@ -1731,15 +1746,55 @@ void OnGUI()
...
+                                    targetData = vesselRadarData.lockedTargetData.targetData;
+                                    radarTarget = true;
+                                }
+                                else if (_irstsEnabled)
                                 {
-                                    distanceToTarget = Vector3.Distance(guardTarget.CoM, ml.MissileReferenceTransform.position);
-                                    Vector3 fireSolution = MissileGuidance.GetAirToAirFireSolution(ml, guardTarget.CoM, guardTarget.Velocity());
+                                    targetData = vesselRadarData.activeIRTarget(null, this);
+                                }
+
```
Comment (SuicidalInsanity) [inline BDArmory/Control/MissileFire.cs:1768]: No support for Scanned but not Locked targets?
Could be simplified down to something like 
```
if (vesselRadarData)
{
    TargetSignatureData targetData = TargetSignatureData.noTarget;
    if (_radarsEnabled)
    {
        if (vesselRadarData.locked)
            targetData = vesselRadarData.lockedTargetData.targetData;
        else
            targetData = vesselRadarData.detectedRadarTarget(guardTarget != null ? guardTarget : null, this);
    }
    else if (_irstsEnabled)
        targetData = vesselRadarData.activeIRTarget(null, this);

    if (targetData.exists)
    {
        distanceToTarget = Vector3.Distance(targetData.predictedPosition, ml.MissileReferenceTransform.position);
        Vector3 fireSolution = MissileGuidance.GetAirToAirFireSolution(ml, targetData.predictedPosition, targetData.velocity);
        Vector3 fsDirection = (fireSolution - ml.MissileReferenceTransform.position).normalized;
        if (vesselRadarData.locked)
            GUIUtils.DrawTextureOnWorldPos(ml.MissileReferenceTransform.position + (distanceToTarget * fsDirection), BDArmorySetup.Instance.greenDotTexture, new Vector2(6, 6), 0);
        else
            GUIUtils.DrawTextureOnWorldPos(ml.MissileReferenceTransform.position + (distanceToTarget * fsDirection), BDArmorySetup.Instance.greenCircleTexture, new Vector2(36, 36), 5);
        GUIUtils.DrawTextureOnWorldPos(ml.MissileReferenceTransform.position + (distanceToTarget * ml.GetForwardTransform()), BDArmorySetup.Instance.dottedLargeGreenCircle, new Vector2(128, 128), 0);
    }
    else
    {
        GUIUtils.DrawTextureOnWorldPos(missile.MissileReferenceTransform.position + (2000 * missile.GetForwardTransform()), BDArmorySetup.Instance.largeGreenCircleTexture, new Vector2(156, 156), 0);
        break;
    }
}
else
    GUIUtils.DrawTextureOnWorldPos(missile.MissileReferenceTransform.position + (2000 * missile.GetForwardTransform()), BDArmorySetup.Instance.largeGreenCircleTexture, new Vector2(156, 156), 0);
break;
```
if you wanted to differentiate between radar/irst detection

**Review by BrettRyland — COMMENTED** [review]
I see `.Velocity()` being used a lot throughout the code.

`.Velocity()` does a sudden switch from surface velocity to orbital velocity at 100km altitude for vessels that are "orbiting", "escaping" or "sub-orbital", which might mess up your calculations if the missiles are being lofted that high (or the targets are that high). So if neither the firing vessel nor the target are in orbit, then you probably want to be using `.srf_velocity` instead.
(I'm fairly sure that the Krakensbane velocity reference frame shifts between the surface and orbital velocities only when the active vessel crosses that line (it was important in Camera Tools), but that shouldn't affect things here since you're not making Krakensbane corrections for non-vessels.)

File: BDArmory/Weapons/ModuleWeapon.cs [hunk 11]
```diff
@@ -5885,7 +5885,7 @@ public void ParseAmmoStats()
                         }
                         else if (eHEType == FillerTypes.Shaped)
                         {
-                            guiAmmoTypeString += StringUtils.Localize("#LOC_BDArmory_Ammo_Shaped") + " ";
+                            guiAmmoTypeString  = StringUtils.Localize("#LOC_BDArmory_Ammo_Shaped") + " ";
```
Comment (BrettRyland) [inline BDArmory/Weapons/ModuleWeapon.cs:5888]: Is this change intentional? It overrides the previous additions to `guiAmmoTypeString`, making (amongst other potential combinations) "cluster semi-armor-piercing nuclear shaped explosive" ammo be described incorrectly.

Comment (BillNyeTheIE) [reply BDArmory/Weapons/ModuleWeapon.cs:5888]: The armor-piercing portion is redundant for shaped charge munitions, also AFAIK you can't have a nuclear shaped charge, at least in terms of the display name, as the if statement above blocks that. I'll have to modify the code to account for cluster munitions though.

File: BDArmory/Control/MissileFire.cs (hunk 7 above)
Comment (BrettRyland) [reply BDArmory/Control/MissileFire.cs:7269]: Using a switch expression, the `unguidedGuidedMissile` parameter here could be written more cleanly as
```csharp
CurrentMissile.TargetingMode switch
{
    MissileBase.TargetingModes.Laser => BDATargetManager.ActiveLasers.Count <= 0,
    MissileBase.TargetingModes.Radar => !(_radarsEnabled || _sonarsEnabled) && !CurrentMissile.radarLOAL,
    MissileBase.TargetingModes.Inertial => !(_radarsEnabled || _sonarsEnabled || _irstsEnabled),
    _ => false
}
```
Similarly for lines 6503—6505.
